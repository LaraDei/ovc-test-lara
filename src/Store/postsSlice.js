import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import  Axios  from 'axios'

export const getPosts = createAsyncThunk(
  'Store/getPosts',
  async (userId, dispatch, getState) => {
    const response = await Axios.get(
      `https://jsonplaceholder.typicode.com/posts?userId=${userId}`
    );
    return response.data
  }
);

export const postsSlice = createSlice({
  name: 'posts',
  initialState:{
    posts: [],
    status: null
  },
  reducers: {
  },
  extraReducers:{
    [getPosts.pending]: (state, action)=>{
      state.status = 'loading'
    },
    [getPosts.fulfilled]: (state, action)=>{
      state.posts = action.payload
      state.status = 'success'
    },
    [getPosts.rejected]: (state, action)=>{
      state.status = 'failed'
    },
  }
});


export default postsSlice.reducer;
