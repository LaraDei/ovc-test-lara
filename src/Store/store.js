import { configureStore } from "@reduxjs/toolkit";
import userReducer from "./usersSlice";
import postsReducer from "./postsSlice";

const rootReducer = { users: userReducer, posts: postsReducer };
export const store = configureStore({
  reducer: rootReducer,
});


