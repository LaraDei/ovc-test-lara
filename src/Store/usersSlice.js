import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import Axios from "axios";

export const getUsers = createAsyncThunk("users/getUsers", async () => {
  const respose = await Axios.get("https://jsonplaceholder.typicode.com/users");
  return respose.data;
});

export const usersSlice = createSlice({
  name: "users",
  initialState: {
    users: [],
    isLoading: false,
    hasError: false,
  },
  reducers: {},
  extraReducers: {
    [getUsers.pending]: (state, action) => {
      state.isLoading = true;
      state.hasError = false;
    },
    [getUsers.fulfilled]: (state, action) => {
      state.users = action.payload;
      state.isLoading = false;
      state.hasError = false;
    },
    [getUsers.rejected]: (state, action) => {
      state.isLoading = false;
      state.hasError = true;
    },
  },
});

export default usersSlice.reducer;
