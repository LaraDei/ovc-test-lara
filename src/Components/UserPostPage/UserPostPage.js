import React, { Component } from "react";
import { connect } from "react-redux";
import { getPosts } from "../../Store/postsSlice";
import TableStyle from "../TableStyle/TableStyle";
import { store } from "../../Store/store";
import { findUser } from "../../helpers";

class UserPostPage extends Component {
  componentDidMount() {
    const { userId } = this.props.match.params;
    this.props.getPosts(userId);
  }

  render() {
    const { userId } = this.props.match.params;
    const { users } = store.getState().users;
    const userName = findUser(users, userId) || { content: "" };
    const { posts } = store.getState().posts;
    const postsData = posts.map((post) => {
      return {
        id: post.id,
        title: post.title,
        body: post.body,
      };
    });
    const headerList = ["Title", "Body"];
    return (
      <div>
        <button
          className="back-btn"
          onClick={() => this.props.history.push('/')}
        >
          BACK
        </button>
        <h2>{userName.name}'s Posts</h2>
        <TableStyle columns={headerList} data={postsData} function={() => {}} />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    posts: state.posts,
  };
};
const mapDispatchToProps = () => {
  return {
    getPosts,
  };
};
export default connect(mapStateToProps, mapDispatchToProps())(UserPostPage);
