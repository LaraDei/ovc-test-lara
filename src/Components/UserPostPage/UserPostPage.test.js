import React from "react";
import ReactDOM from "react-dom";
import "../../setupTests"
import { shallow } from 'enzyme';
import renderer from "react-test-renderer";
import UserPostPage from "./UserPostPage";
import { Provider } from "react-redux";
import { store } from "../../Store/store";
import sinon from 'sinon';


describe("UserPostPage component", () => {
  const props = {
    users: [
      {
        id: 1,
        name: "testName",
        email: "testEmal",
        city: "testCity",
        company: "testCompany",
      },
      {
        id: 2,
        name: "testName",
        email: "testEmal",
        city: "testCity",
        company: "testCompany",
      },
    ],
    posts: [
      {
        id: 1,
        userId: 2,
        title: "testTitle",
        body: "testBody",
      },
      {
        id: 2,
        userId: 1,
        title: "testTitle",
        body: "testBody",
      },
      {
        id: 3,
        userId: 2,
        title: "testTitle",
        body: "testBody",
      },
    ],
  };
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Provider store={store}>
        <UserPostPage
          users={props.users}
          match={{ params: { id: 1 }, isExact: true, path: "", url: "" }}
        />
      </Provider>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it("renders by default", () => {
    const wrapper = renderer
      .create(
        <Provider store={store}>
          <UserPostPage
            users={props.users}
            match={{ params: { id: 1 }, isExact: true, path: "", url: "" }}
          />
        </Provider>
      )
      .toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});
