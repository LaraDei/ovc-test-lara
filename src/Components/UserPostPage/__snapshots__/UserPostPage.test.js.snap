// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPostPage component it renders by default 1`] = `
<div>
  <button
    className="back-btn"
    onClick={[Function]}
  >
    BACK
  </button>
  <h2>
    's Posts
  </h2>
  <div
    className="table-container"
  >
    <table
      className="table"
    >
      <thead>
        <tr>
          <th>
            Title
          </th>
          <th>
            Body
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;

exports[`UserPostPage component renders by default 1`] = `
<div>
  <button
    className="back-btn"
    onClick={[Function]}
  >
    BACK
  </button>
  <h2>
    's Posts
  </h2>
  <div
    className="table-container"
  >
    <table
      className="table"
    >
      <thead>
        <tr>
          <th>
            Title
          </th>
          <th>
            Body
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
